### PR TITLE
Add `fsspec` support for `to_json`, `to_csv`, and `to_parquet`

### PR DESCRIPTION
**Pull request by alvarobartt**
Hi to whoever is reading this! 🤗 (Most likely @mariosasko)

## What's in this PR?

This PR replaces the `open` from Python with `fsspec.open` and adds the argument `storage_options` for the methods `to_json`, `to_csv`, and `to_parquet`, to allow users to export any 🤗`Dataset` into a file in a file-system as requested at #6086.

## What's missing in this PR?

As per `to_json`, `to_csv`, and `to_parquet` docstrings for the recently included `storage_options` arg, I've scoped it to 2.15.0, so we should check that before merging in case we want to scope that for 2.14.2 instead.

Additionally, should we also add `fsspec` support for the `from_csv`, `from_json`, and `from_parquet` methods? If you want me to do so @mariosasko just let me know and I'll create another PR to support that too!

### CODE REVIEW

**Comment by alvarobartt** [comment]
Hi @mariosasko and/or @lhoestq, friendly pining you guys here! Let me know if there's anything else to be included within this PR, thanks! 🤗

**Comment by alvarobartt** [comment]
Hi here @lhoestq @mariosasko I just realised this PR is still open, should we close it in case this is something not to include within `datasets`, or should we merge? Let me know whatever you decide 🤗

File: src/datasets/arrow_dataset.py [hunk 1]
```diff
@@ -4709,13 +4709,15 @@ def to_csv(
...
         num_proc: Optional[int] = None,
+        storage_options: Optional[dict] = None,
         **to_csv_kwargs,
     ) -> int:
         """Exports the dataset to csv
 
         Args:
             path_or_buf (`PathLike` or `FileOrBuffer`):
-                Either a path to a file or a BinaryIO.
+                Either a path to a file (e.g. `file.csv`), a remote URI (e.g. `s3://my-bucket/file.csv`),
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:4719]: taking the opportunity to advertise `hf://` paths 🙃 

```suggestion
                Either a path to a file (e.g. `file.csv`), a remote URI (e.g. `hf://datasets/username/my_dataset_name/data.csv`),
```

File: src/datasets/arrow_dataset.py [hunk 2]
```diff
@@ -4824,13 +4837,15 @@ def to_json(
...
         num_proc: Optional[int] = None,
+        storage_options: Optional[dict] = None,
         **to_json_kwargs,
     ) -> int:
         """Export the dataset to JSON Lines or JSON.
 
         Args:
             path_or_buf (`PathLike` or `FileOrBuffer`):
-                Either a path to a file or a BinaryIO.
+                Either a path to a file (e.g. `file.json`), a remote URI (e.g. `s3://my-bucket/file.json`),
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:4847]:
```suggestion
                Either a path to a file (e.g. `file.json`), a remote URI (e.g. `hf://datasets/username/my_dataset_name/data.json`),
```

File: src/datasets/arrow_dataset.py [hunk 3]
```diff
@@ -4907,16 +4933,22 @@ def to_parquet(
...
         batch_size: Optional[int] = None,
+        storage_options: Optional[dict] = None,
         **parquet_writer_kwargs,
     ) -> int:
         """Exports the dataset to parquet
 
         Args:
             path_or_buf (`PathLike` or `FileOrBuffer`):
-                Either a path to a file or a BinaryIO.
+                Either a path to a file (e.g. `file.parquet`), a remote URI (e.g. `s3://my-bucket/file.parquet`),
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:4943]:
```suggestion
                Either a path to a file (e.g. `file.parquet`), a remote URI (e.g. `hf://datasets/username/my_dataset_name/data.parquet`),
```

File: src/datasets/arrow_dataset.py [hunk 4]
```diff
@@ -4724,6 +4726,10 @@ def to_csv(
                 use multiprocessing. `batch_size` in this case defaults to
                 `datasets.config.DEFAULT_MAX_BATCH_SIZE` but feel free to make it 5x or 10x of the default
                 value if you have sufficient compute power.
+            storage_options (`dict`, *optional*):
+                Key/value pairs to be passed on to the file-system backend, if any.
+
+                <Added version="2.15.0"/>
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:4732]:
```suggestion
                <Added version="2.19.0"/>
```

File: src/datasets/arrow_dataset.py [hunk 5]
```diff
@@ -4839,6 +4854,10 @@ def to_json(
...
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:4860]: fix added version

```suggestion
                <Added version="2.19.0"/>
```

File: src/datasets/arrow_dataset.py [hunk 6]
```diff
@@ -4907,16 +4933,22 @@ def to_parquet(
...
-                Either a path to a file or a BinaryIO.
+                Either a path to a file (e.g. `file.parquet`), a remote URI (e.g. `s3://my-bucket/file.parquet`),
+                or a BinaryIO, where the dataset will be saved to in the specified format.
             batch_size (`int`, *optional*):
                 Size of the batch to load in memory and write at once.
                 Defaults to `datasets.config.DEFAULT_MAX_BATCH_SIZE`.
+            storage_options (`dict`, *optional*):
+                Key/value pairs to be passed on to the file-system backend, if any.
+
+                <Added version="2.15.0"/>
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:4951]:
```suggestion
                <Added version="2.19.0"/>
```

**Review by lhoestq — COMMENTED** [review]
Thanks for the ping ! It looks good to me, I just added a few suggestions before merging:

**Comment by albertvillanova** [comment]
Thanks @alvarobartt. 

I am linking this PR to the corresponding issue (on the right column, under "Development") and closing the issue.

For future contributions, please add to the PR description the word "fix" followed by the issue number, e.g.:
```
Fix #6086.
```

**Comment by alvarobartt** [comment]
> Thanks @alvarobartt.
> 
> I am linking this PR to the corresponding issue (on the right column, under "Development") and closing the issue.
> 
> For future contributions, please add to the PR description the word "fix" followed by the issue number, e.g.:
> 
> ```
> Fix #6086.
> ```
> 
> * I have edited the PR description to add this.

Hi @albertvillanova, fair, I missed that, thanks for the edit and the heads up!

**Pull request merged**